etcd rollback tool panic

I tried to do a rollback on my own test cluster and got a panic.  I will attach the data-dir tarball.  The script that performs the migration goes from 2.2.1->2.3.7->3.0.14->2.3.7, and it is this last step that barfs.  Script is forthcoming in a PR, but there's a snapshot at https://gist.github.com/mml/42e57c45b333f8b1c8c29bb162394e7d

```
 Performing etcd3 -> etcd2 rollback
 + /usr/local/bin/rollback --data-dir /var/etcd/data
 2016-11-10 03:18:07.226168 C | membership: error updating attributes of unknown member cf7d617fc919f571
 panic: error updating attributes of unknown member cf7d617fc919f571

 goroutine 1 [running]:
 panic(0xae1600, 0xc8211be610)
         /usr/local/go/src/runtime/panic.go:481 +0x3e6
 k8s.io/kubernetes/vendor/github.com/coreos/pkg/capnslog.(*PackageLogger).Panicf(0xc820138ac0, 0xe5b340, 0x2e, 0xc8219112e8, 0x1, 0x1)
         /go/src/k8s.io/kubernetes/vendor/github.com/coreos/pkg/capnslog/pkg_logger.go:75 +0x191
 k8s.io/kubernetes/vendor/github.com/coreos/etcd/etcdserver/membership.(*RaftCluster).UpdateAttributes(0xc8213c2b40, 0xcf7d617fc919f571, 0xc8211be520, 0xb, 0xc8211c8f80, 0x1, 0x4)
         /go/src/k8s.io/kubernetes/vendor/github.com/coreos/etcd/etcdserver/membership/cluster.go:331 +0x2e5
 k8s.io/kubernetes/vendor/github.com/coreos/etcd/etcdserver.(*applierV2store).Put(0xc8202a2be0, 0xc8214d11e0, 0x0, 0x0, 0x0, 0x0, 0x0)
         /go/src/k8s.io/kubernetes/vendor/github.com/coreos/etcd/etcdserver/apply_v2.go:83 +0x765
 main.applyRequest(0xc8214d11e0, 0x7f4c439473c0, 0xc8202a2be0)
         /go/src/k8s.io/kubernetes/cluster/images/etcd/rollback/rollback.go:321 +0x1cf
 main.rebuild(0x7ffc57679f18, 0xe, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
         /go/src/k8s.io/kubernetes/cluster/images/etcd/rollback/rollback.go:296 +0xe8d
 main.main()
         /go/src/k8s.io/kubernetes/cluster/images/etcd/rollback/rollback.go:103 +0x70c
```

@wojtek-t @hongchaodeng

>
panic: error updating attributes of unknown member cf7d617fc919f571

This seems that it's not even about the real data, but about some metadata of the cluster.

@hongchaodeng @xiang90 - any thoughts on it?

--- a/cluster/images/etcd/migrate-if-needed.sh
+++ b/cluster/images/etcd/migrate-if-needed.sh
@@ -222,6 +222,12 @@ if [ "${CURRENT_STORAGE}" = "etcd3" -a "${TARGET_STORAGE}" = "etcd2" ]; then
     echo "etcd3 -> etcd2 downgrade is supported only between 3.0.x and 2.3.7"
     return 0
   fi
+  echo "Backup and remove all existing v2 data"
+  ROLLBACK_BACKUP_DIR="${DATA_DIRECTORY}.bak"
+  rm -rf "${ROLLBACK_BACKUP_DIR}"
+  mkdir -p "${ROLLBACK_BACKUP_DIR}"
+  cp -r "${DATA_DIRECTORY}" "${ROLLBACK_BACKUP_DIR}"
+  rm -rf "${DATA_DIRECTORY}"/member/snap/*.snap
   echo "Performing etcd3 -> etcd2 rollback"
   ${ROLLBACK} --data-dir "${DATA_DIRECTORY}"
   if [ "$?" -ne "0" ]; then

